parquet-column/src/test/java/org/apache/parquet/column/values/bloom/Bloom.java

```diff
@@ -0,0 +1,412 @@
+  }
+
+  /**
+   * Construct the bloom filter with given bit set, it is used
+   * when reconstruct bloom filter from parquet file.
+   * @param bitset The given bitset to construct bloom filter.
+   * @param hash The hash strategy bloom filter apply.
+   * @param algorithm The algorithm of bloom filter.
+   */
+  public Bloom(byte[] bitset, HASH hash, ALGORITHM algorithm) {
+    this.bitset = bitset;
+    this.intBuffer = ByteBuffer.wrap(bitset).asIntBuffer();
+    this.hash = hash;
+    this.algorithm = algorithm;
+  }
+
+  /**
+   * Construct the bloom filter with given bit set.
+   * @param input The given bitset to construct bloom filter.
+   * @param hash The hash strategy bloom filter apply.
+   * @param algorithm The algorithm of bloom filter.
+   */
+  public Bloom(ByteBuffer input, HASH hash, ALGORITHM algorithm) {
+    if (input.isDirect()){
+      bitset = new byte[input.remaining()];
+      input.get(bitset);
+    } else {
+      this.bitset = input.array();
+    }
+    this.intBuffer = ByteBuffer.wrap(bitset).asIntBuffer();
+    this.hash = hash;
+    this.algorithm = algorithm;
+  }
+
+  /**
+   * Create a new bitset for bloom filter, at least 256 bits will be create.
+   * @param numBytes number of bytes for bit set.
+   */
+  private void initBitset(int numBytes) {
+    if (numBytes < BYTES_PER_BUCKET) {
+      numBytes = BYTES_PER_BUCKET;
+    }
+
+    // Get next power of 2 if it is not power of 2.
+    if ((numBytes & (numBytes - 1)) != 0) {
+      numBytes = Integer.highestOneBit(numBytes) << 1;
+    }
+
+    if (numBytes > ParquetProperties.DEFAULT_MAXIMUM_BLOOM_FILTER_SIZE) {
+      numBytes = ParquetProperties.DEFAULT_MAXIMUM_BLOOM_FILTER_SIZE;
+    }
+
+    this.bitset = new byte[numBytes];
+    this.intBuffer = ByteBuffer.wrap(bitset).asIntBuffer();
+  }
+
+  /**
+   * Write bloom filter to output stream. A bloom filter structure should include
+   * bitset length, hash strategy, algorithm, and bitset.
+   * @param out output stream to write
+   */
+  public void writeTo(OutputStream out) throws IOException {
+    Preconditions.checkArgument(bitset.length > 0, "Bloom filter bitset length should be larger than 0");
+
+    // Write number of bytes of bitset.
+    out.write(BytesUtils.intToBytes(bitset.length));
+
+    // Write hash strategy
+    out.write(BytesUtils.intToBytes(this.hash.ordinal()));
+
+    // Write algorithm
+    out.write(BytesUtils.intToBytes(this.algorithm.ordinal()));
+
+    // Write bitset
+    out.write(bitset);
+  }
+
+  private int[] setMask(int key) {
+    // The block based algorithm needs 8 odd SALT values to calculate eight index
+    // of bit to set, one bit in 32-bit word.
+    final int SALT[] = {0x47b6137b, 0x44974d91, 0x8824ad5b, 0xa2b7289d,
+      0x705495c7, 0x2df1424b, 0x9efc4947, 0x5c6bfb31};
+
+    int mask[] = new int[8];
+
+    for (int i = 0; i < 8; ++i) {
+      mask[i] = key * SALT[i];
+    }
+
+    for (int i = 0; i < 8; ++i) {
+      mask[i] = mask[i] >>> 27;
```

Comment:
For switching between little-endian and big-endian, I think you can subtract mask[i] from 31 after this operation.

Reply:
No need to switch endianness now.